library/command/command/src/main/java/org/quiltmc/qsl/command/impl/client/ClientCommandInternals.java

```diff
@@ -0,0 +1,289 @@
+		}
+
+		return executeHelp(Iterables.getLast(nodes).getNode(), context);
+	}
+
+	private static int executeHelp(CommandNode<QuiltClientCommandSource> startNode, CommandContext<QuiltClientCommandSource> context) {
+		Map<CommandNode<QuiltClientCommandSource>, String> commands = DISPATCHER.getSmartUsage(startNode, context.getSource());
+
+		for (String command : commands.values()) {
+			context.getSource().sendFeedback(new LiteralText(PREFIX + command));
+		}
+
+		return commands.size();
+	}
+
+	private static LiteralArgumentBuilder<QuiltClientCommandSource> createRunCommand() {
+		LiteralArgumentBuilder<QuiltClientCommandSource> runCommand = literal("run");
+		for (CommandNode<QuiltClientCommandSource> node : ClientCommandManager.DISPATCHER.getRoot().getChildren()) {
+			runCommand.then(node);
+		}
+		return runCommand;
+	}
+
+	public static void addCommands(CommandDispatcher<QuiltClientCommandSource> target, QuiltClientCommandSource source) {
+		Map<CommandNode<QuiltClientCommandSource>, CommandNode<QuiltClientCommandSource>> originalToCopy = new HashMap<>();
+		originalToCopy.put(DISPATCHER.getRoot(), target.getRoot());
+		copyChildren(DISPATCHER.getRoot(), target.getRoot(), source, originalToCopy);
+	}
+
+	/**
+	 * Copies the child commands from origin to target, filtered by {@code child.canUse(source)}.
+	 * Mimics vanilla's CommandManager.makeTreeForSource.
+	 *
+	 * @param origin         the source command node
+	 * @param target         the target command node
+	 * @param source         the command source
+	 * @param originalToCopy a mutable map from original command nodes to their copies, used for redirects;
+	 *                       should contain a mapping from origin to target
+	 */
+	private static void copyChildren(
+			CommandNode<QuiltClientCommandSource> origin,
+			CommandNode<QuiltClientCommandSource> target,
+			QuiltClientCommandSource source,
+			Map<CommandNode<QuiltClientCommandSource>, CommandNode<QuiltClientCommandSource>> originalToCopy
+	) {
+		for (CommandNode<QuiltClientCommandSource> child : origin.getChildren()) {
+			if (!child.canUse(source)) continue;
+
+			if (target.getChild(child.getName()) != null) {
+				continue;
+			}
+
+			ArgumentBuilder<QuiltClientCommandSource, ?> builder = child.createBuilder();
+
+			// Reset the unnecessary non-completion stuff from the builder
+			builder.requires(s -> true); // This is checked with the if check above.
+
+			if (builder.getCommand() != null) {
+				builder.executes(DUMMY_COMMAND);
+			}
+
+			// Set up redirects
+			if (builder.getRedirect() != null) {
+				builder.redirect(originalToCopy.get(builder.getRedirect()));
+			}
+
+			CommandNode<QuiltClientCommandSource> result = builder.build();
+			originalToCopy.put(child, result);
+			target.addChild(result);
+
+			if (!child.getChildren().isEmpty()) {
+				copyChildren(child, result, source, originalToCopy);
+			}
+		}
+	}
+
+	private static final Command<QuiltClientCommandSource> DUMMY_COMMAND = ctx -> 0;
```

Comment:
Putting this with the other `private static final` fields would help with clarity.

Reply:
Done.